### PR TITLE
Xeno Tiebreaker: King & Xeno psychic points

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

Complementary PR to tank, waiting on the wound state sprites from splinter and probbaly should change some action icons
Xeno psychic points are for the entire hive and can be used to maybe make a xeno version of req
The psychic points are gained from corrupting power generators as queen or shrike

The king is summoned by the queen through a King Pod when the hive gets enough Psychic points

New Abilities:
Psychic whisper: Distorts marine vision and aim when activated, has a pretty effect and last a a nice while
Gravity crush: Minor wind up after which a 3x3 area is heavily damaged, good for cades
Hive summon: After a wind-up with telegraphs the entire hive is teleported to the king

Shrike and Queen Ability:
Corrupt generator
## Changelog
:cl:
add: Xeno King
add: Xeno Psychic hive points
admin: Logging for hive summon in game log
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+	if(!do_after(X, 15 SECONDS, FALSE, X, BUSY_ICON_HOSTILE))
+		for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
+			sister.remove_filter("summonoutline")
+		return fail_activate()
+
+	for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
+		sister.remove_filter("summonoutline")
+		sister.forceMove(get_turf(X))
+	log_attack("[key_name(owner)] has summoned hive ([length(allxenos)] Xenos) in [AREACOORD(owner)]")
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:142]:
```suggestion
	log_game("[key_name(owner)] has begun summoning hive in [AREACOORD(owner)]")
	xeno_message("King: \The [owner] has begun a psychic summon in <b>[get_area(owner)]</b>!", 3, X.hivenumber)
	var/list/allxenos = X.hive.get_all_xenos()
	for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
		sister.add_filter("summonoutline", 2, list("type" = "outline", "size" = 1, "color" = COLOR_VIOLET))

	if(!do_after(X, 15 SECONDS, FALSE, X, BUSY_ICON_HOSTILE))
		for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
			sister.remove_filter("summonoutline")
		return fail_activate()

	for(var/mob/living/carbon/xenomorph/sister as() in allxenos)
		sister.remove_filter("summonoutline")
		sister.forceMove(get_turf(X))
	log_game("[key_name(owner)] has summoned hive ([length(allxenos)] Xenos) in [AREACOORD(owner)]")
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+/datum/action/xeno_action/activable/whisper/use_ability(atom/A)
+	var/mob/living/carbon/xenomorph/X = owner
+
+	succeed_activate()
+	add_cooldown()
+
+	X.whispering = TRUE
+	X.update_icon()
+	playsound(X.loc, 'sound/voice/king_xeno_whispers.ogg', 125, 0)
+	X.visible_message("<span class='xenohighdanger'>\The [X] opens it's jaw and silently screeches!</span>")
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:31]:
```suggestion
	X.visible_message("<span class='xenohighdanger'>\The [X] opens its jaw and silently screeches!</span>")
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	action_icon_state = "screech"
+	mechanics_text = "Increases the localized gravity in an area and crushes things."
+	ability_name = "Gravity crush"
+	plasma_cost = 100
+	cooldown_timer = 30 SECONDS
+	keybind_flags = XACT_KEYBIND_USE_ABILITY//mark
+	keybind_signal = COMSIG_XENOABILITY_SCREECH //tivi mark
+
+/datum/action/xeno_action/activable/area_crush/on_cooldown_finish()
+	to_chat(owner, "<span class='warning'>Our psychic aura restores itself. We are ready to whisper again.</span>")
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:77]:
```suggestion
	to_chat(owner, "<span class='warning'>Our psychic aura restores itself. We are ready to area crush again.</span>")
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	if(!iscarbon(target))
+		return ..()
+	if(!can_use_ability(target, override_flags = XACT_IGNORE_SELECTED_ABILITY))
+		return ..()
+	return TRUE
+
+/datum/action/xeno_action/psychic_summon
+	name = "Psychic Summon"
+	action_icon_state = "screech"
+	mechanics_text = "Summons all xenos in a hive to the callers location."
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:106]:
```suggestion
	mechanics_text = "Summons all xenos in a hive to the caller's location, uses all plasma to activate."
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/king.dm [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+/mob/living/carbon/xenomorph/king/handle_special_wound_states(severity)
+	. = ..()
+	if(whispering)
+		return "king_wounded_screeching"
+
+
+///resin pod that creates the king xeno after a delay
+/obj/structure/resin/king_pod
+	name = "Psychic echo chamber"
+	desc = "A resin monolith used to channel the Queen Mothers energy to summon elder xenos to a battlefield."
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/king/king.dm:53]:
```suggestion
	desc = "A resin monolith used to channel the Queen Mother's energy to summon elder xenomorphs to the battlefield."
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/king.dm [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+/mob/living/carbon/xenomorph/king/handle_special_wound_states(severity)
+	. = ..()
+	if(whispering)
+		return "king_wounded_screeching"
+
+
+///resin pod that creates the king xeno after a delay
+/obj/structure/resin/king_pod
+	name = "Psychic echo chamber"
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/king/king.dm:52]:
```suggestion
	name = "psychic echo chamber"
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/king.dm [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	kong.offer_mob()
+
+///When the king mob is offered and then accepted this proc ejects the king and does announcements
+/obj/structure/resin/king_pod/proc/on_king_occupied(mob/occupied)
+	UnregisterSignal(occupied, COMSIG_MOB_LOGIN)
+	occupied.forceMove(loc)
+	desc += " This one has already been used."
+	var/myarea = get_area(src)
+	priority_announce("Warning: Psychic anomaly signature in [myarea] has spiked and begun to move.", "TGMC Intel Division")
+	xeno_message("<span class='xenoannounce'>[occupied] has awakened at [myarea]. Praise the queen mother!</span>", 3 ,ownerhive)
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/king/king.dm:90]:
```suggestion
	xeno_message("<span class='xenoannounce'>[occupied] has awakened at [myarea]. Praise the Queen Mother!</span>", 3 ,ownerhive)
```

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 8]
```diff
@@ -706,3 +706,54 @@
...
+	var/obj/machinery/power/geothermal/gen = A
+	if(!do_after(owner, 10 SECONDS, TRUE, gen, BUSY_ICON_HOSTILE))
+		return fail_activate()
+	var/mob/living/carbon/xenomorph/X = owner
+	gen.corrupted = X.hivenumber
+	gen.update_icon()
+	gen.start_processing()
+
+/datum/action/xeno_action/summon_king
+	name = "Summon a Xeno King"
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:734]:
```suggestion
	name = "Summon Xenomorph King"
```

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 9]
```diff
@@ -706,3 +706,54 @@
...
+		return
+	var/mob/living/carbon/xenomorph/X = owner
+	if(SSpoints.xeno_points_by_hive["[X.hivenumber]"] <= REQUIRED_POINTS_FOR_KING_SUMMON)
+		if(!silent)
+			to_chat(owner, "<span class='warning'>Our hive does not have enough psychic energy to summon a King, we need [REQUIRED_POINTS_FOR_KING_SUMMON-SSpoints.xeno_points_by_hive["[X.hivenumber]"]] more.</span>")
+		return FALSE
+
+/datum/action/xeno_action/summon_king/action_activate()
+	var/mob/living/carbon/xenomorph/X = owner
+	to_chat(X, "<span class='xenonotice'>We begin constructing a psychic echo chamber for the queen mother...</span>")
```
Comment (XSlayer300) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:752]:
```suggestion
	to_chat(X, "<span class='xenonotice'>We begin constructing a psychic echo chamber for the Queen Mother...</span>")
```

**Comment by XSlayer300** [comment]
Two things besides those:

First, can you make sure that the Xeno king is designated as a "hive leader", to ensure the hivemind collapse countdown isn't triggered by the death of the queen or shrike.

Next, can you make the Xeno King timelocked or atleast give priority to players over an X amount of playtime as a xeno?

**Comment by neohenry** [comment]
I sense that psychic points will expand xenomorph gameplay a lot.

**Comment by neohenry** [comment]
Counterplays against xenomorphs generating psychic points:

Marines mortaring engineering since it is easy to get coordinates of generators by just opening webmap.

**Comment by TiviPlus** [comment]
> Counterplays against xenomorphs generating psychic points:
> 
> Marines mortaring engineering since it is easy to get coordinates of generators by just opening webmap.

As addressed in discord, corruption is removed by welder

**Comment by TiviPlus** [comment]
> Two things besides those:
> 
> First, can you make sure that the Xeno king is designated as a "hive leader", to ensure the hivemind collapse countdown isn't triggered by the death of the queen or shrike.
> 
> Next, can you make the Xeno King timelocked or atleast give priority to players over an X amount of playtime as a xeno?

A) Its meant to end stalemates not prolong them
B) No, same as tank isnt locked

**Comment by XSlayer300** [comment]
Next thing, gravity crush clears weeds, intentional? It also does self-damage to xenos (and yourself) too, intentional as well?

**Comment by Pariah919** [comment]
>Complementary PR to tank, waiting on the wound state sprites from splinter and probbaly should change some action icons
Xeno psychic points are for the entire hive and can be used to maybe make a xeno version of req
The psychic points are gained from corrupting power generators as queen or shrike
>
if you think this is comparable to the tank you're having a bruh. If the tank is supposed to be the comparable to this nerf king hp to 500 hp and make him unable to heal at all, or make the tank use something other than req points to be bought, as you're effectively crippling marines in every part of the game while xenomorphs just sit around and get rewarded for existing basically.

The tank requires marines to pause all requisition progression and every single utility/power item they get from this, while psychic points (unless you tie something I can't think of considering xenomorphs are so shallow in scale and gameplay) doesn't really cost you anything. This thing only requires xenomorphs to have basic map control.

>The king is summoned by the queen through a King Pod when the hive gets enough Psychic points
>
No comment

>New Abilities:
Psychic whisper: Distorts marine vision and aim when activated, has a pretty effect and last a a nice while
Gravity crush: Minor wind up after which a 3x3 area is heavily damaged, good for cases
>Hive summon: After a wind-up with telegraphs the entire hive is teleported to the king
Hive summon seems silly to me, gravity crush is meh game design to me but whatever it kills cades

>Shrike and Queen Ability:
>Corrupt generator
>
replace this with some other way to generate poins. Add actual generators that actually provide points and map them in with sdmm, rather than just giving every generator the ability to be corrupted because in maps like ice colony you've literally broken the game

**Comment by Pariah919** [comment]
Adding on to the above point-

The tank requires more than a 2 man crew to use it effectively actually, as a large number of our maps are pretty tight corridored, so you need someone to clear for it, which adds on to the crew count to make the tank actually function effectively, and you need someone to help that guy, then that guy needs a medic to help that guy.. And if any of these people fuck up the tank is effectively a useless garbage meme that people laugh at because it can't do anything, costs 750 points (the cost of 6! 6! SADARs) which I could use to outfit out a gamer squad of ungas who I know are actually competent at the game with useful equipment that'll actually break a stalemate if they're good at the game.

The skill disparity between both """""""stalemate"""""""" breakers is really big.

**Comment by neohenry** [comment]
Gravity crush is pretty devastating.

It's heavy explosive, thus the same as OB. It can make marines horizontal pretty easily, not to mention that gravity crush has a 30 seconds cooldown. If the tank is facing against a king xeno & both tank users and king xeno have the same skills, king xeno will win when the king xeno manages to get out of tank's firing lane and drop gravity crush multiple times (with that 3x3 OB, tank will be paper compared to king xeno).

**Comment by Pariah919** [comment]
Infact if you want to make this equivalent to the tank. Do this.

Make it have 200/225/250/275 HP. So it dies to an AP rocket or whatever.
Give it 55/60/65/70 armor.
Nerf whisper to 5 seconds and it requires LOS give it minute long cooldown.
Gravity crush has a 2 minute cooldown.
Now it actually requires its team to win rather than just winning on its own! It's still better than the tank since it requires significantly less coordination, but if it was actually equal to the tank it'd be way worse than this.
Though personally I dislike the way this idea was thought out.

**Comment by AdamElTablawy** [comment]
It doesn't need any nerfs. You're meant to lose if the xenos roll this thing out.

**Comment by SkeletalElite** [comment]
@AdamElTablawy 

Pretty sure mortaring generators doesn't work. you have to use a welder on the generator to uncorrupt it I think

**Comment by InterroLouis** [comment]
Are you not meant to lose if marines roll out the tank?

**Comment by AdamElTablawy** [comment]
@InterroLouis You are, yeah

**Comment by Kuro-ai** [comment]
> @InterroLouis You are, yeah
> 
> Just saying the tank on the whole being less powerful makes a fair bit of sense since it's easier to secure one (750 points as compared to an uninterrupted hour+ for the king with no queen deaths)

but it's not easier to secure,750 points seems easy to get until you realize that it's gonna get dragged on since marines need to buy other shit and that there are better stuff to buy in req than the tank,the tank is a meme and king xenos needs a nerf from it's current "insta win" strenght

**Comment by MLGTASTICa** [comment]
> > @InterroLouis You are, yeah
> > Just saying the tank on the whole being less powerful makes a fair bit of sense since it's easier to secure one (750 points as compared to an uninterrupted hour+ for the king with no queen deaths)
> 
> but it's not easier to secure,750 points seems easy to get until you realize that it's gonna get dragged on since marines need to buy other shit and that there are better stuff to buy in req than the tank,the tank is a meme and king xenos needs a nerf from it's current "insta win" strenght

If marines ever bothered to coordinate pushing and securing chokepoints , getting a tank would be actually easy. King xeno is a tie breaker and so it the tank , they are not meant to be balanced at all.

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 10]
```diff
@@ -0,0 +1,164 @@
+/datum/action/xeno_action/activable/whisper
+	name = "Whisper"
+	action_icon_state = "emit_neurogas2"
```
Comment (Hughgent) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:3]: to be changed I presume

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+		return ..()
+	if(get_dist(target, owner) > 5)
+		return ..()
+	if(!can_use_ability(target, override_flags = XACT_IGNORE_SELECTED_ABILITY))
+		return ..()
+	return TRUE
+
+/datum/action/xeno_action/activable/area_crush
+	name = "Gravity Crush"
+	action_icon_state = "fortify"
```
Comment (Hughgent) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:71]: placeholder

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+/datum/action/xeno_action/activable/psychic_summon/on_cooldown_finish()
+	to_chat(owner, "<span class='warning'>The hives power swells. We may summon our sisters again.</span>")
+	return ..()
+
+/datum/action/xeno_action/psychic_summon/can_use_action(silent, override_flags)
+	. = ..()
+	var/mob/living/carbon/xenomorph/X = owner
+	if(length(X.hive.get_all_xenos()) <= 1)
+		if(!silent)
+			to_chat(owner, "<span class='notice'>We have no hive to call.</span>")
```
Comment (Hughgent) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:140]:
```suggestion
			to_chat(owner, "<span class='notice'>We have no hive to call. We are alone on our throne of nothing.</span>")
```

File: code/modules/mob/living/carbon/xenomorph/castes/king/king.dm [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+		if(XENO_UPGRADE_THREE)
+			name = "[hive.prefix]Ancient Emperor ([nicknumber])" //Ancient
+
+	real_name = name
+	if(mind)
+		mind.name = name
+
+/mob/living/carbon/xenomorph/king/handle_special_state()
+	if(whispering)
+		icon_state = "King Screeching"
```
Comment (Hughgent) [inline code/modules/mob/living/carbon/xenomorph/castes/king/king.dm:40]: are you sure this is good `icon_state` naming? 

"king_screeching"

**Review by Hughgent — APPROVED** [review]
I think this is overall good.

**Review by TheBimmer231 — APPROVED** [review]

**Comment by AndroidMK2** [comment]
Hope that unfunny coders keep this change .

**Comment by David-Stormwell** [comment]
noooooo, my king was slain by the fearsome monster known as github-actions bot.... what ever shall we do now....

**Pull request closed without merging**